### PR TITLE
LCOW - Change platform parser directive to FROM statement flag

### PR DESCRIPTION
**Pull request by lowenna**
This removes the `# platform=....` parser directive, and adds `--platform` as a flag to the `FROM` statement instead. This is one of the action items from the F2F in SF in August 2017, and the document described in #34617. It's a direct follow-on to https://github.com/moby/moby/pull/34859 which is now merged in master

image

### CODE REVIEW

**Comment by tianon** [comment]
@stevvooe this is one of the sort of things you had in mind when you created https://github.com/containerd/containerd/pull/1403, right?

**Comment by stevvooe** [comment]
@tianon Yes. We need to massage that package to meet these use cases.

**Comment by jhowardmsft** [comment]
Rebased on latest iteration of https://github.com/moby/moby/pull/34859, and taken account of merging of #35429

**Comment by jhowardmsft** [comment]
Rebased on #34859 again to keep fresh

**Comment by jhowardmsft** [comment]
@johnstep @tonistiigi @stevvooe PTAL. I've now rebased it after #34859 was merged and tidied it up. Ready for review.

**Comment by jhowardmsft** [comment]
@tonistiigi Fixed CI issue and rebased again. PTAL.

**Comment by jhowardmsft** [comment]
@tiborvass Can we move this back to code review? If you see the linked issue (https://github.com/moby/moby/issues/34617), you can see this design was thrashed out at length and agreed upon last year with at least 6 other maintainers....

**Comment by jhowardmsft** [comment]
Oops, I meant @vdemeester in previous comment.

**Comment by thaJeztah** [comment]
ping @stevvooe PTAL: discussing this in the maintainers session, and while the change in this PR (code-wise) looks ok, we want to confirm if making this feature (previously undocumented) is ok.

**Comment by dnephin** [comment]
Does this sound right for the default behaviour ?

case  | result
-----------|---------
Base image is single platform, that platform is supported by the engine | build using that platform
Base image has multiple platforms, one is supported by the engine | build using that platform
Base image has multiple platforms, none are supported by the engine | error platform isn't supported
Base image has multiple platforms, more than one is supported by the engine | use the engine native platform

In the case of LCOW, why should `FROM alpine:latest` fail to build on windows? If the engine is able to figure out that the base image is linux/amd64 why not just proceed with the build? 

The other case would be something like `FROM microsoft/dotnet`. That should continue to build the same way after LCOW with these defaults.

I think the only thing that is missing from this PR is the second case. Is that right?

File: builder/dockerfile/dispatchers.go
Comment (dnephin) [inline builder/dockerfile/dispatchers.go]: This is difficult to follow. I think this would be more clear:

```go
switch {
case stagePlatform != "":
    return stagePlatform
case d.builder.options.Platform != "":
    return d.builder.options.Platform
default:
    return runtime.GOOS
}
```

File: builder/dockerfile/builder.go
Comment (dnephin) [inline builder/dockerfile/builder.go]: Not a blocker for this PR, but we shouldn't really be passing around a string for OS. This paltform should be parsed into a `spec.Platform` in the API layer, and we should use that object everywhere instead of a string.

File: builder/dockerfile/instructions/parse.go
Comment (dnephin) [inline builder/dockerfile/instructions/parse.go]: Since we're adding this here, why not store the `specPlatform` instead of just `OS`?

File: daemon/image_history.go
Comment (dnephin) [inline daemon/image_history.go]: How does this relate to the rest of the PR?

File: daemon/images.go
Comment (dnephin) [inline daemon/images.go]: How does this relate to the rest of the PR?

**Review by dnephin — CHANGES_REQUESTED** [review]

File: daemon/image_history.go
Comment (jhowardmsft) [reply daemon/image_history.go]: It's a safety check to ensure that we don't panic on L39 where the layerstores array is indexed by the operating system. Strictly it's not related to this PR, it should have been in https://github.com/moby/moby/pull/34859, for which this PR is a direct follow up. Happy to move it out to a separate PR if you prefer.

File: daemon/images.go
Comment (jhowardmsft) [reply daemon/images.go]: It's a safety check to ensure that we don't panic on L277 where the layerstores array is indexed by the operating system. Strictly it's not related to this PR, it should have been in https://github.com/moby/moby/pull/34859, for which this PR is a direct follow up. Happy to move it out to a separate PR if you prefer.

File: builder/dockerfile/instructions/parse.go
Comment (jhowardmsft) [reply builder/dockerfile/instructions/parse.go]: If you see https://github.com/moby/moby/issues/34617, and the google doc linked to in the description, that's the write up of the F2F in August last year. Specifically the piece about multi-arch. The daemon is not currently multi-arch throughout, that's a follow-on that someone could do at a later time, but is out of scope for LCOW. Currently for LCOW, I have extended it for multi-OS, and have hence tried to keep it very clear in the code base when there's a "platform" (implying OS+Arch+...) as opposed to an OS. Hence why just "OperatingSystem".

Comment (dnephin) [reply builder/dockerfile/instructions/parse.go]: My request is not to make the daemon multi-arch.

If someone specifies `platform=linux/arm` on an amd64 system I believe this is going to silently succeed. I don't think that's correct.

We should be use platform and the `spec.Platform` type everywhere. Passing strings around is error prone and more likely to hide bugs.

Comment (dnephin) [reply builder/dockerfile/instructions/parse.go]: I think you may also need the entire platform spec to do the proposed default of checking against the manifest list.

Comment (jhowardmsft) [reply builder/dockerfile/instructions/parse.go]: Actually no, not so. in `parseFrom` it calls system.ValidatePlatform which will error out.


```
	// Based on https://github.com/moby/moby/pull/34642#issuecomment-330375350, do
	// not support anything except operating system.
	if platform.Architecture != "" {
		return fmt.Errorf("invalid platform architecture %q", platform.Architecture)
	}
```

Hence you get the following pulling arm on amd64:
```
PS E:\docker\build\arm> docker build .
Sending build context to Docker daemon  2.048kB
Error response from daemon: Dockerfile parse error line 1: invalid platform "linux/arm" on FROM
PS E:\docker\build\arm> type .\Dockerfile
FROM --platform=linux/arm busybox
RUN echo above should fail
PS E:\docker\build\arm>
```

Comment (jhowardmsft) [reply builder/dockerfile/instructions/parse.go]: Hence why deliberately using operating system as a string here until someone wants to make the entire daemon multi-arch as well.

File: builder/dockerfile/dispatchers.go
Comment (jhowardmsft) [reply builder/dockerfile/dispatchers.go]: OK, can simplify, but your logic above isn't quite correct. Priority is the d.builder.options.Platform

Comment (dnephin) [reply builder/dockerfile/dispatchers.go]: I agree the `options.Platform` should be the highest priority, so I'm glad we're making that change.

However that's not what your code does, or what it says in the document:

> If both the API and a `FROM` statement both contain a platform, the one in the FROM statement will be used.

File: builder/dockerfile/instructions/parse.go
Comment (dnephin) [reply builder/dockerfile/instructions/parse.go]: Why discard useless information that we are going to need?

File: builder/dockerfile/dispatchers.go
Comment (jhowardmsft) [reply builder/dockerfile/dispatchers.go]: Actually, not thinking this morning. You were right the first time. Will push an update shortly

File: builder/dockerfile/instructions/parse.go
Comment (jhowardmsft) [reply builder/dockerfile/instructions/parse.go]: It isn't needed until the daemon is fully multi-arch aware, which as I say, is out of scope for LCOW. See distribution\pull_v2* function `filterManifests` for example. It only matches manifests that match `runtime.GOARCH`.

**Comment by jhowardmsft** [comment]
@tonistiigi
> The thing left to fix is the default behavior that happens when no flag is specified but multiple platforms are supported.

@dnephin
> I think the only thing that is missing from this PR is the second case. Is that right?

You guys are essentially talking about the same thing. That is a separate topic which was talked about in the F2F where at least Tonis was present. See https://github.com/moby/moby/issues/34617 and the google doc specifically. At the very end of the doc, it mentions this, but it was left optional and still needs more discussion for implementation in another PR. That said, I don't disagree with you, just it doesn't need to be part of this PR. I can work on that separately.

> Dockerfiles are supposed to work on multiple machines with different configurations and platforms. 99% of the cases the stage is only valid for a single platform so we shouldn't force users to edit a dockerfile when moving from linux to lcow system.

There is absolutely no requirement to edit the dockerfile. That's why the API/CLI supports the platform flag too. I do this all the time to build existing pre-canned dockerfiles which are based on Linux images. Of course, if the image is already present, that flag is not required and it would "just work" anyway.

**Comment by tonistiigi** [comment]
> I don't disagree with you, just it doesn't need to be part of this PR. I can work on that separately.

I agree it is separate, but this PR really makes this behavior usable for the first time. The previous logic is in code but nobody really knows about it. So it could be separate but then I'd like the priority fix first. Otherwise, we are adding a new flag and then changing the behavior of what that flag really means after that.

>  That's why the API/CLI supports the platform flag too. I do this all the time to build existing pre-canned dockerfiles which are based on Linux images. 

That is the problem. You make your Dockerfiles and run them with special configuration, then I take this file, don't know your special flags and they don't work for me, making a bad user experience. Ideally, we shouldn't have the override in API at all for build but ok for me to leave it as a fallback.

**Comment by jhowardmsft** [comment]
> So it could be separate but then I'd like the priority fix first.

Remember though that LCOW is still only enabled under experimental. That's why I don't believe it needs to be part of this PR. I can commit to putting something together once the functionality in this PR is merged.

**Comment by jhowardmsft** [comment]
@tonistiigi @dnephin PTAL at https://github.com/moby/moby/pull/36418, specifically the last commit (currently 9fffa6b51b91c55fa78ed187d0ce3bfee8ffc2f5) which is the follow up to both this PR and https://github.com/moby/moby/pull/36327 to do the autoselection of Linux if it's available and usable. It's not ready for review, but I believe does what you are asking.

**Comment by jhowardmsft** [comment]
Rebased.

**Comment by jhowardmsft** [comment]
Rebased due to the new images package under daemon/ was introduced

**Comment by jhowardmsft** [comment]
Ping @tonistiigi @dnephin. PTAL along with https://github.com/moby/moby/pull/36418 which addressed the last part of the previous discussion.

File: builder/dockerfile/instructions/parse.go
Comment (tonistiigi) [inline builder/dockerfile/instructions/parse.go]: This shouldn't be in the parser package. The point of this package is to just convert from strings into typed structures. Validation should be based on the structures parser returns.

Comment (jhowardmsft) [reply builder/dockerfile/instructions/parse.go]: @tonistiigi Fixed, update pushed.

File: builder/dockerfile/builder.go
Comment (dnephin) [inline builder/dockerfile/builder.go]: This seems like a strange place to validate the platform.

Why is this necessary? Don't we already do this as the first step in `dispatch` ?

If it's still necessary then maybe `dispatchDockerfileWithCancellation` is a better place to do this validation?

File: builder/dockerfile/dispatchers.go
Comment (dnephin) [inline builder/dockerfile/dispatchers.go]: This variable will only contain `OS` right? should be renamed to avoid confusion, or use the full platform struct.

File: builder/dockerfile/dispatchers.go
Comment (dnephin) [inline builder/dockerfile/dispatchers.go]: This variable will only contain `OS` right? should be renamed to avoid confusion, or use the full platform struct.

File: builder/dockerfile/dispatchers.go
Comment (dnephin) [inline builder/dockerfile/dispatchers.go]: This variable will only contain `OS` right? should be renamed to avoid confusion, or use the full platform struct.

File: builder/dockerfile/dispatchers.go
Comment (jhowardmsft) [reply builder/dockerfile/dispatchers.go]: Fair. Update being pushed shortly.

File: builder/dockerfile/dispatchers.go
Comment (jhowardmsft) [reply builder/dockerfile/dispatchers.go]: Fair. Update being pushed shortly.

File: builder/dockerfile/dispatchers.go
Comment (jhowardmsft) [reply builder/dockerfile/dispatchers.go]: Fair. Update being pushed shortly.

File: builder/dockerfile/builder.go
Comment (jhowardmsft) [reply builder/dockerfile/builder.go]: It put it here based on https://github.com/moby/moby/pull/35089#discussion_r173266198. This is the next highest place in the call-stack based on that comment, once all the dockerfile has been parsed. By validating here, it's done before any build step is executed, rather than failing after one or more build steps have been executed.

File: builder/dockerfile/dispatchers.go
Comment (jhowardmsft) [reply builder/dockerfile/dispatchers.go]: Pushed now.

File: builder/dockerfile/builder.go
Comment (jhowardmsft) [reply builder/dockerfile/builder.go]: Note this is also consistent with where all the all the instructions are validated immediately before it - again as opposed to running one or more steps, only to subsequently fail.

Comment (dnephin) [reply builder/dockerfile/builder.go]: That comment just says to not do it in the parser package, not necessarily to do it here.

I still don't think this is the correct place to do this validation. Stages should be validated as part of dispatch  not randomly in some other unrelated function.

Comment (jhowardmsft) [reply builder/dockerfile/builder.go]: @dnephin OK, moved to dispatch

**Review by dnephin — APPROVED** [review]
LGTM

**Review by johnstep — APPROVED** [review]
LGTM

File: builder/dockerfile/instructions/parse.go
Comment (johnstep) [inline builder/dockerfile/instructions/parse.go]: Nit: I am not a fan of this Hungarian-style notation, preferring `platform` or `platformFlag`. I realize there is a precedent for the `fl` prefix in the builder code, so it's okay, for consistency, but I think it should be cleaned up eventually. :)

**Pull request merged**
